## Byte Pair Encoding
This notebook demonstrates how to use byte pair encoding (BPE) for tokenization.

## Install tiktoken
Install the tiktoken library, which efficiently implements the BPE algorithm.

In [ ]:
pip install tiktoken

## Initialize and Use the BPE Tokenizer
Set up the BPE tokenizer from tiktoken and demonstrate encoding and decoding.

In [32]:
import tiktoken

tokenizer = tiktoken.get_encoding("gpt2")
# encode the sample text
sample_text = "Hello, do you like tea? <|endoftext|> I hope you do!"
encoded = tokenizer.encode(sample_text, allowed_special={"<|endoftext|>"})
print("Encoded:", encoded)

# decode the tokens back to original text
decoded = tokenizer.decode(encoded)
print("Decoded:", decoded)

Encoded: [15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 314, 2911, 345, 466, 0]
Decoded: Hello, do you like tea? <|endoftext|> I hope you do!


## Special Tokens in tiktoken
The <|endoftext|> token is assigned the largest token ID (50256) in a vocabulary of 50,257 tokens.